src/util/pcp.cpp

```diff
@@ -0,0 +1,524 @@
+//!  Mapping nonce (12 bytes).
+constexpr size_t PCP_MAP_NONCE_OFS = 0;
+//!  Protocol (1 byte).
+constexpr size_t PCP_MAP_PROTOCOL_OFS = 12;
+//!  Internal port for mapping (2 bytes).
+constexpr size_t PCP_MAP_INTERNAL_PORT_OFS = 16;
+//!  Suggested external port (request), assigned external port (response) (2 bytes).
+constexpr size_t PCP_MAP_EXTERNAL_PORT_OFS = 18;
+//!  Suggested external IP (request), assigned external IP (response) (16 bytes).
+constexpr size_t PCP_MAP_EXTERNAL_IP_OFS = 20;
+
+//! Result code representing success (RFC6887 7.4), shared with NAT-PMP.
+constexpr uint8_t PCP_RESULT_SUCCESS = NATPMP_RESULT_SUCCESS;
+//! Result code representing lack of resources (RFC6887 7.4).
+constexpr uint8_t PCP_RESULT_NO_RESOURCES = 8;
+
+//! Mapping of PCP result code to string (RFC6887 7.4). Result codes <=2 match NAT-PMP.
+const std::map<uint8_t, std::string> PCP_RESULT_STR{
+    {0,  "SUCCESS"},
+    {1,  "UNSUPP_VERSION"},
+    {2,  "NOT_AUTHORIZED"},
+    {3,  "MALFORMED_REQUEST"},
+    {4,  "UNSUPP_OPCODE"},
+    {5,  "UNSUPP_OPTION"},
+    {6,  "MALFORMED_OPTION"},
+    {7,  "NETWORK_FAILURE"},
+    {8,  "NO_RESOURCES"},
+    {9,  "UNSUPP_PROTOCOL"},
+    {10, "USER_EX_QUOTA"},
+    {11, "CANNOT_PROVIDE_EXTERNAL"},
+    {12, "ADDRESS_MISMATCH"},
+    {13, "EXCESSIVE_REMOTE_PEER"},
+};
+
+//! Return human-readable string from NATPMP result code.
+std::string NATPMPResultString(uint8_t result_code)
+{
+    auto result_i = NATPMP_RESULT_STR.find(result_code);
+    return strprintf("%s (code %d)", result_i == NATPMP_RESULT_STR.end() ? "(unknown)" : result_i->second,  result_code);
+}
+
+//! Return human-readable string from PCP result code.
+std::string PCPResultString(uint8_t result_code)
+{
+    auto result_i = PCP_RESULT_STR.find(result_code);
+    return strprintf("%s (code %d)", result_i == PCP_RESULT_STR.end() ? "(unknown)" : result_i->second,  result_code);
+}
+
+//! Wrap address in IPv6 according to RFC6887. wrapped_addr needs to be able to store 16 bytes.
+[[nodiscard]] bool PCPWrapAddress(Span<uint8_t> wrapped_addr, const CNetAddr &addr)
+{
+    Assume(wrapped_addr.size() == ADDR_IPV6_SIZE);
+    if (addr.IsIPv4()) {
+        struct in_addr addr4;
+        if (!addr.GetInAddr(&addr4)) return false;
+        // Section 5: "When the address field holds an IPv4 address, an IPv4-mapped IPv6 address [RFC4291] is used (::ffff:0:0/96)."
+        std::memcpy(wrapped_addr.data(), IPV4_IN_IPV6_PREFIX.data(), IPV4_IN_IPV6_PREFIX.size());
+        std::memcpy(wrapped_addr.data() + IPV4_IN_IPV6_PREFIX.size(), &addr4, ADDR_IPV4_SIZE);
+        return true;
+    } else if (addr.IsIPv6()) {
+        struct in6_addr addr6;
+        if (!addr.GetIn6Addr(&addr6)) return false;
+        std::memcpy(wrapped_addr.data(), &addr6, ADDR_IPV6_SIZE);
+        return true;
+    } else {
+        return false;
+    }
+}
+
+//! Unwrap PCP-encoded address according to RFC6887.
+CNetAddr PCPUnwrapAddress(Span<const uint8_t> wrapped_addr)
```

Comment:
In cbb053c5f6fa63b08fe8fb200b143cca64fc0626 "net: Add PCP and NATPMP implementation"

`CNetAddr` has a `SetLegacyIPv6` method which does the same thing. Could use that instead of duplicating implementation.

Reply:
Maybe the other way around, replace the `CNetAddr` implementation with what you wrote here. But that sounds like a followup.

Reply:
Wouldn't that have the same conceptual issue, but the other way around? eg locking the bitcoin implementation to RFC6887 specifics. 

It didn't feel right, it's different specifications coming from a different place that somehow in a parallel evolution ended up similarly.

Now looking at `SetLegacyIPv6` i see the differences: 
- there's the TORV2 case. It just marks the addess as invalid, which it would be anyway, so *that* wouldn't be problematic here.
- there's also the `NET_INTERNAL` case, which makes no sense in this context.

i think it's the second one that made me doubt the most-i have no idea what else we'd want to encode into addresses IPv6 in the future, but it shouldn't interfere with the PCP implementation.